source/blog/2015-05-08-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,216 @@
+
+```ruby
+def match(expected, actual)
+  actual.match expected
+end
+```
+
+Now obviously, one could extend the capabilites by providing other objects that provided a
+definition of match, but that's besides the point.
+
+The new definition of the match matcher's match method is thus:
+
+```ruby
+def match(expected, actual)
+  return true if values_match?(expected, actual)
+  return false unless can_safely_call_match?(expected, actual)
+  actual.match(expected)
+end
+```
+
+The key important difference is that call to `values_match?` which is the core implementation
+of the new matching protocol in RSpec. The list of matchers which now reference the values_match?
+method are (as of RSpec expectations 3.3.0 development, not including mocks):
+
+* `change`
+* `contain_exactly`
+* `have_attributes`
+* `include`
+* `match`
+* `output`
+* `raise_error`
+* `start_or_end_with`
+* `throw_symbol`
+* `yield`
+
+Let's take a look at what `values_match?` does.
+
+`values_match?` is the RSpec match definition that allows matchers, and other objects, to compose.
+The implementation looks like this:
+
+```ruby
+def self.values_match?(expected, actual)
+  if Hash === actual
+    return hashes_match?(expected, actual) if Hash === expected
+  elsif Array === expected && Enumerable === actual && !(Struct === actual)
+    return arrays_match?(expected, actual.to_a)
+  end
+
+  return true if expected == actual
+
+  begin
+    expected === actual
+  rescue ArgumentError
+    # Some objects, like 0-arg lambdas on 1.9+, raise
+    # ArgumentError for `expected === actual`.
+    false
+  end
+end
+```
+
+The first thing to note is the special casing for Hashes and Arrays. Basically,
+what those methods do is recursively apply `values_match?` to each of the
+elements in the various collection types as you'd expect. Arrays by scanning
+elements and hashes by matching keys and values.
+
+The next thing the method does is return `true` if expected === actual. An easy
+way to demonstrate this is back in our IRB session:
+
+```
+>> o.expect(3).to o.match(3)
+=> true
+```
+
+What we've done here is we've constructed a both a target and match matcher with the value 3.
+When the match matcher hits `values_match?` it sees that 3 is not an array or a hash, and so
+checks `expected == actual`, which in this case is `3 == 3`, which is true.
+
+The next line checks `expected === actual`. This is where the power of RSpec's new composability
+really kicks in. The reason for this is that the `BaseMatcher` class includes a module called
+`Composable`. The composable matcher defines `===` on matchers thusly:
+
+```rspec
+def ===(value)
+  matches?(value)
+end
+```
+
+This means that `expected` in the expression `expected === actual` can be another RSpec matcher,
+a regex, or any other object that responds to `===`. This means, for example, that one can trivially
+match a lambda by doing `expect(obj).to match(lambda { |actual| ... })`. Given that all RSpec matchers
```

Comment:
Unfortunately, not all RSpec matchers implement `===`.  All _built in_ matchers do, and all matches defined using the DSL do...but custom matchers defined as stand-alone classes do not unless the user has defined `===` or included `Composable`.